benchmarks/distributed/rpc/parameter_server/trainers/DdpTrainer.py

```diff
@@ -0,0 +1,167 @@
+        Args:
+            rank (int): worker rank
+            trainer_count (int): count of trainer in the world
+            process_group (object): distributed process group
+            use_cuda_rpc (bool): indicator for CUDA RPC
+            server_rref (object): remote reference to the server
+            backend (string): distributed communication backend
+            epochs (int): epoch count for training
+        """
+        super().__init__(rank)
+        self.rank = rank
+        self.trainer_count = trainer_count
+        self.process_group = process_group
+        self.use_cuda_rpc = use_cuda_rpc
+        self.server_rref = server_rref
+        self.backend = backend
+        self.epochs = epochs
+
+    @staticmethod
+    def hook(state, bucket):
+        r"""
+        ddp communication hook that uses the current backend allreduce
+        implementation.
+        Args:
+            state (object): maintains state during the training process
+            bucket (object): gradient bucket
+        """
+        cref = state.cref
+        tensors = [bucket.get_tensor() / state.process_group.size()]
+        key = state.get_key(bucket.get_index())
+        cref.record_hook_fut_start(key, f"{cref.backend}_allreduce")
+        fut = state.process_group.allreduce(tensors).get_future()
+
+        def callback(fut):
+            cref.record_hook_fut_end(key)
+            return fut.wait()
+
+        return fut.then(callback)
+
+    def get_hook(self):
+        r"""
+        returns DdpTrainer.hook
+        """
+        return DdpTrainer.hook
+
+    def create_ddp_model(self, model):
+        r"""
+        creates a ddp_model and hook_state objects, registers ddp_model communication hook,
+        and returns the ddp_model and hook_state.
+        Args:
+            model (object): neural network model
+        """
+        ddp_model = DDP(
+            model, device_ids=[self.rank], process_group=self.process_group
+        )
+        hook_state = self.HookState(self, self.process_group)
+        ddp_model.register_comm_hook(hook_state, self.get_hook())
+        return ddp_model, hook_state
+
+    def epoch_key(self, epoch, index):
+        r"""
+        returns an encoded key that represents the current epoch and
+        iteration index.
+        Args:
+            epoch (int): epoch index
+            index (int): iteration index
+        """
+        return f"{epoch},{index}"
+
+    def preprocess_data(self, data):
+        r"""
+        moves the data from CPU to GPU.
+        Args:
+            data (list): training examples
+        """
+        for i in range(len(data)):
+            data[i][0] = data[i][0].cuda(self.rank)
+            data[i][1] = data[i][1].cuda(self.rank)
+        return data
+
+    def iteration_step(self, ddp_model, criterion, optimizer, hook_state, epoch, index, batch):
+        r"""
+        performs an iteration of training.
+        Args:
+            ddp_model (object): distributed data parallel model
```

Comment:
`ddp_model` should be an `nn.Module` instance?

Reply:
okay.